gcsio/src/main/java/com/google/cloud/hadoop/gcsio/MetricsPublisher.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.hadoop.gcsio;
+
+import static io.opencensus.common.Duration.fromMillis;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.auth.Credentials;
+import com.google.cloud.hadoop.util.CredentialAdapter;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import io.opencensus.common.Scope;
+import io.opencensus.contrib.grpc.metrics.RpcViews;
+import io.opencensus.exporter.stats.stackdriver.StackdriverStatsConfiguration;
+import io.opencensus.exporter.stats.stackdriver.StackdriverStatsExporter;
+import io.opencensus.stats.Aggregation;
+import io.opencensus.stats.Aggregation.Distribution;
+import io.opencensus.stats.BucketBoundaries;
+import io.opencensus.stats.Measure.MeasureLong;
+import io.opencensus.stats.Stats;
+import io.opencensus.stats.StatsRecorder;
+import io.opencensus.stats.View;
+import io.opencensus.stats.View.Name;
+import io.opencensus.stats.ViewManager;
+import io.opencensus.tags.TagContext;
+import io.opencensus.tags.TagContextBuilder;
+import io.opencensus.tags.TagKey;
+import io.opencensus.tags.TagMetadata;
+import io.opencensus.tags.TagMetadata.TagTtl;
+import io.opencensus.tags.TagValue;
+import io.opencensus.tags.Tagger;
+import io.opencensus.tags.Tags;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.concurrent.GuardedBy;
+
+/** Interface for exposing utility methods to publish metrics via open-census api-spec. */
+interface MetricsPublisher {
```

Comment:
Can we consolidate OpenCensus metrics publishing with Hadoop's Statistics infra, so we do not duplicate code that doing roughly the same, just exposes data differently?

Reply:
ping

Reply:
There are a lot of fine grain metrics (message level latencies, status of each rpc, retries), some of which don't make sense to be aggregated. 

IMO, Hadoop stats and Opencensus metrics are operating at different levels of abstractions. 
If we consolidate Opencensus metrics and Hadoop Stats, it would be difficult to extend more metrics without co-ordinating between multiple layers, complicating the code-base. Spark metrics [architecture](https://db-blog.web.cern.ch/blog/luca-canali/2019-02-performance-dashboard-apache-spark) does encourage publishing directly to different sinks for metrics which are not viable to be aggregated at the master.

Reply:
Makes sense, thanks for explanation.